File: src/market_maker_sim/book.py
"""A price-time-priority limit order book matching engine.

Semantics (standard continuous double auction, as on most equity exchanges):

* **Price priority**: an incoming order always trades against the best-priced
  opposite orders first (highest bid / lowest ask).
* **Time priority (FIFO)**: within a price level, orders execute in arrival
  order. Arrival order is defined by an engine-assigned sequence number, not
  wall-clock time, so it is total and unambiguous.
* **Execution price**: every fill prints at the *resting* order's limit price.
  An aggressive order that crosses several levels gets price improvement level
  by level ("walking the book").
* **Marketable limit orders** match immediately for whatever is available
  inside their limit, then rest the remainder at their limit price.
* **Market orders** match against everything available; any remainder is
  discarded (emitted as a ``Cancelled`` event with reason ``unfilled_market``)
  — market orders never rest.

Data structures
---------------
Each side is a ``dict[price_ticks, deque[RestingOrder]]`` (the FIFO queue per
level) plus a binary heap of level prices for O(log n) best-price access. The
heap is maintained *lazily*: prices whose level has emptied are popped on the
next peek. Cancellation is *eager* (the order is removed from its level's deque
immediately, O(level size)). Production engines usually prefer lazy tombstoning
with intrusive linked lists for O(1) cancels; eager removal is chosen here
because it keeps a much stronger invariant — every order present in a queue is
live — which makes the matching loop simpler to reason about and to test, and
level sizes in this simulation are small.
"""

from __future__ import annotations

import heapq
from collections import deque

from .orders import Accepted, BookEvent, Cancelled, Fill, RestingOrder, Side

__all__ = ["LimitOrderBook"]


class LimitOrderBook:
    def __init__(self) -> None:
        # price level -> FIFO queue of live resting orders
        self._levels: dict[Side, dict[int, deque[RestingOrder]]] = {
            Side.BUY: {},
            Side.SELL: {},
        }
        # Lazy heaps of level prices. Bid prices are stored negated so that
        # heapq's min-heap yields the highest bid first.
        self._price_heap: dict[Side, list[int]] = {Side.BUY: [], Side.SELL: []}
        self._orders: dict[int, RestingOrder] = {}  # live orders by id, for cancel
        self._next_order_id = 1
        self._next_entry_seq = 1

    # ------------------------------------------------------------------ views

    def best_bid(self) -> int | None:
        return self._best_price(Side.BUY)

    def best_ask(self) -> int | None:
        return self._best_price(Side.SELL)

    def mid(self) -> float | None:
        """Mid price in ticks (may be half-integer). None if either side is empty."""
        bid, ask = self.best_bid(), self.best_ask()
        if bid is None or ask is None:
            return None
        return (bid + ask) / 2

    def spread(self) -> int | None:
        bid, ask = self.best_bid(), self.best_ask()
        if bid is None or ask is None:
            return None
        return ask - bid

    def depth(self, side: Side, n_levels: int | None = None) -> list[tuple[int, int]]:
        """Aggregate (price, total_qty) per level, best first."""
        levels = self._levels[side]
        prices = sorted(levels, reverse=(side is Side.BUY))
        if n_levels is not None:
            prices = prices[:n_levels]
        return [(p, sum(o.qty for o in levels[p])) for p in prices]

    def get_order(self, order_id: int) -> RestingOrder | None:
        return self._orders.get(order_id)

    def order_count(self) -> int:
        return len(self._orders)

    # ---------------------------------------------------------------- actions

    def submit_limit(
        self, owner: str, side: Side, price: int, qty: int, time: float
    ) -> tuple[int, list[BookEvent]]:
        """Submit a limit order. Returns (order_id, events).

        The order first matches against any crossing liquidity (it is the
        taker), then the remainder rests at ``price``.
        """
        self._validate(price=price, qty=qty)
        order_id = self._new_order_id()
        events: list[BookEvent] = []
        remaining = self._match(
            taker_id=order_id,
            taker_owner=owner,
            side=side,
            qty=qty,
            limit_price=price,
            time=time,
            events=events,
        )
        if remaining > 0:
            self._rest(order_id, owner, side, price, remaining)
            events.append(
                Accepted(time=time, order_id=order_id, owner=owner, side=side,
                         price=price, qty=remaining)
            )
        return order_id, events

    def submit_market(
        self, owner: str, side: Side, qty: int, time: float
    ) -> tuple[int, list[BookEvent]]:
        """Submit a market order. Any quantity that cannot be matched is
        discarded (reported as Cancelled/unfilled_market)."""
        self._validate(qty=qty)
        order_id = self._new_order_id()
        events: list[BookEvent] = []
        remaining = self._match(
            taker_id=order_id,
            taker_owner=owner,
            side=side,
            qty=qty,
            limit_price=None,
            time=time,
            events=events,
        )
        if remaining > 0:
            events.append(
                Cancelled(time=time, order_id=order_id, owner=owner, side=side,
                          price=None, qty=remaining, reason="unfilled_market")
            )
        return order_id, events

    def cancel(self, order_id: int, time: float) -> Cancelled | None:
        """Cancel a resting order. Returns the event, or None if the order is
        unknown or already fully filled/cancelled (a no-op, as on real venues
        where cancels race with fills)."""
        order = self._orders.pop(order_id, None)
        if order is None:
            return None
        level = self._levels[order.side][order.price]
        level.remove(order)  # identity-based; O(level size)
        if not level:
            del self._levels[order.side][order.price]
        return Cancelled(time=time, order_id=order_id, owner=order.owner,
                         side=order.side, price=order.price, qty=order.qty,
                         reason="user")

    # --------------------------------------------------------------- matching

    def _match(
        self,
        taker_id: int,
        taker_owner: str,
        side: Side,
        qty: int,
        limit_price: int | None,
        time: float,
        events: list[BookEvent],
    ) -> int:
        """Match an incoming order against the opposite side. Mutates the book,
        appends Fill events, and returns the unmatched remainder."""
        pre_trade_mid = self.mid()
        opposite = side.opposite
        remaining = qty
        while remaining > 0:
            best = self._best_price(opposite)
            if best is None:
                break
            # Crossing test, written sign-uniformly: a BUY crosses when its
            # limit >= best ask; a SELL crosses when its limit <= best bid.
            if limit_price is not None and side.sign * (limit_price - best) < 0:
                break
            queue = self._levels[opposite][best]
            while remaining > 0 and queue:
                maker = queue[0]
                traded = min(remaining, maker.qty)
                maker.qty -= traded
                remaining -= traded
                events.append(
                    Fill(time=time, price=maker.price, qty=traded,
                         taker_order_id=taker_id, maker_order_id=maker.order_id,
                         taker_owner=taker_owner, maker_owner=maker.owner,
                         taker_side=side, pre_trade_mid=pre_trade_mid)
                )
                if maker.qty == 0:
                    queue.popleft()
                    del self._orders[maker.order_id]
            if not queue:
                del self._levels[opposite][best]
        return remaining

    # ---------------------------------------------------------------- helpers

    def _rest(self, order_id: int, owner: str, side: Side, price: int, qty: int) -> None:
        order = RestingOrder(order_id=order_id, owner=owner, side=side,
                             price=price, qty=qty, entry_seq=self._next_entry_seq)
        self._next_entry_seq += 1
        levels = self._levels[side]
        if price not in levels:
            levels[price] = deque()
            key = -price if side is Side.BUY else price
            heapq.heappush(self._price_heap[side], key)
        levels[price].append(order)
        self._orders[order_id] = order

    def _best_price(self, side: Side) -> int | None:
        """Best live price on a side, cleaning stale heap entries lazily."""
        heap = self._price_heap[side]
        levels = self._levels[side]
        while heap:
            price = -heap[0] if side is Side.BUY else heap[0]
            if price in levels:
                return price
            heapq.heappop(heap)
        return None

    def _new_order_id(self) -> int:
        order_id = self._next_order_id
        self._next_order_id += 1
        return order_id

    @staticmethod
    def _validate(qty: int, price: int | None = None) -> None:
        if not isinstance(qty, int) or qty <= 0:
            raise ValueError(f"qty must be a positive integer, got {qty!r}")
        if price is not None and not isinstance(price, int):
            raise ValueError(f"price must be an integer tick count, got {price!r}")

    # ------------------------------------------------------------- invariants

    def check_invariants(self) -> None:
        """Assert internal consistency. Used by tests after every operation;
        never called on the hot path."""
        bid, ask = self.best_bid(), self.best_ask()
        assert bid is None or ask is None or bid < ask, (
            f"book is crossed: bid {bid} >= ask {ask}"
        )
        seen: set[int] = set()
        for side in Side:
            for price, queue in self._levels[side].items():
                assert queue, f"empty level {price} left in book"
                seqs = [o.entry_seq for o in queue]
                assert seqs == sorted(seqs), f"FIFO order violated at level {price}"
                for order in queue:
                    assert order.qty > 0, f"zero-qty order {order.order_id} in book"
                    assert order.price == price and order.side == side
                    assert self._orders.get(order.order_id) is order
                    seen.add(order.order_id)
        assert seen == set(self._orders), "order index out of sync with levels"

File: ocaml/book.ml
(** A price-time-priority limit order book matching engine.

    Semantics are those of a standard continuous double auction:

    - {b Price priority}: an incoming order trades against the best-priced
      opposite orders first (highest bid / lowest ask).
    - {b Time priority (FIFO)}: within a price level, orders execute in arrival
      order, defined by an engine-assigned sequence number.
    - {b Execution price}: every fill prints at the {i resting} order's limit
      price, so an aggressive order that crosses several levels gets price
      improvement level by level ("walking the book").
    - {b Marketable limit orders} match immediately for whatever is available
      inside their limit, then rest the remainder.
    - {b Market orders} never rest; any unmatched remainder is discarded and
      reported as [Cancelled] with reason ["unfilled_market"].

    {2 Data structures}

    Each side is a map from price tick to that level's FIFO queue, held as a
    list with the oldest order at the head. The Python implementation of this
    engine pairs a hash map of levels with a lazily-cleaned binary heap of
    prices; here a balanced-tree [Map] subsumes both, since [max_binding] and
    [min_binding] give the best bid and ask directly in [O(log n)]. Empty levels
    are deleted eagerly, so every price in the map is live and no stale-entry
    cleanup is needed.

    Cancellation removes the order from its level immediately rather than
    tombstoning it. That is [O(level size)], but it buys a much stronger
    invariant, which is that every order reachable from a queue is live, and
    level sizes in this simulation are small. *)

open Orders

module IntMap = Map.Make (Int)

type t = {
  mutable bids : resting_order list IntMap.t;
  mutable asks : resting_order list IntMap.t;
  mutable orders : resting_order IntMap.t;  (** live orders by id, for cancel *)
  mutable next_order_id : int;
  mutable next_entry_seq : int;
}

let create () =
  {
    bids = IntMap.empty;
    asks = IntMap.empty;
    orders = IntMap.empty;
    next_order_id = 1;
    next_entry_seq = 1;
  }

let levels book = function Buy -> book.bids | Sell -> book.asks

let set_levels book side m =
  match side with Buy -> book.bids <- m | Sell -> book.asks <- m

(* ------------------------------------------------------------------ views *)

let best_bid book =
  match IntMap.max_binding_opt book.bids with Some (p, _) -> Some p | None -> None

let best_ask book =
  match IntMap.min_binding_opt book.asks with Some (p, _) -> Some p | None -> None

let best_price book = function Buy -> best_bid book | Sell -> best_ask book

(** Mid price in ticks; may be a half-integer, hence [float]. [None] if either
    side is empty. *)
let mid book =
  match (best_bid book, best_ask book) with
  | Some b, Some a -> Some (float_of_int (b + a) /. 2.)
  | _ -> None

let spread book =
  match (best_bid book, best_ask book) with
  | Some b, Some a -> Some (a - b)
  | _ -> None

let level_qty q = List.fold_left (fun acc o -> acc + o.qty) 0 q

(** Aggregate [(price, total_qty)] per level, best first. *)
let depth book side n_levels =
  let bindings = IntMap.bindings (levels book side) in
  (* [IntMap.bindings] is ascending by price: best is the last for bids, the
     first for asks. *)
  let ordered = match side with Buy -> List.rev bindings | Sell -> bindings in
  let ordered =
    match n_levels with
    | None -> ordered
    | Some k -> List.filteri (fun i _ -> i < k) ordered
  in
  List.map (fun (p, q) -> (p, level_qty q)) ordered

let get_order book order_id = IntMap.find_opt order_id book.orders
let order_count book = IntMap.cardinal book.orders

(* --------------------------------------------------------------- matching *)

let new_order_id book =
  let id = book.next_order_id in
  book.next_order_id <- id + 1;
  id

let rest_order book ~order_id ~owner ~side ~price ~qty =
  let order =
    { order_id; owner; side; price; qty; entry_seq = book.next_entry_seq }
  in
  book.next_entry_seq <- book.next_entry_seq + 1;
  let m = levels book side in
  let q = Option.value (IntMap.find_opt price m) ~default:[] in
  (* Append to the tail: newest order has lowest time priority. *)
  set_levels book side (IntMap.add price (q @ [ order ]) m);
  book.orders <- IntMap.add order_id order book.orders

(** Match an incoming order against the opposite side. Mutates the book and
    returns [(remaining, events)] with events in chronological order.

    [limit_price] is [None] for market orders, which cross at any price. *)
let match_incoming book ~taker_id ~taker_owner ~side ~qty ~limit_price ~time =
  let pre_trade_mid = mid book in
  let opp = opposite side in
  let remaining = ref qty in
  let events = ref [] in
  let keep_going = ref true in
  while !keep_going && !remaining > 0 do
    match best_price book opp with
    | None -> keep_going := false
    | Some best ->
        (* Crossing test, written sign-uniformly: a Buy crosses when its limit
           is at or above the best ask, a Sell when at or below the best bid. *)
        let crosses =
          match limit_price with
          | None -> true
          | Some lp -> sign side * (lp - best) >= 0
        in
        if not crosses then keep_going := false
        else begin
          let queue = IntMap.find best (levels book opp) in
          (* Walk this level's FIFO, consuming makers until the incoming order
             is filled or the level is exhausted. Returns what is left of the
             queue plus the fills generated, newest first. *)
          let rec walk q acc =
            if !remaining <= 0 then (q, acc)
            else
              match q with
              | [] -> ([], acc)
              | maker :: rest ->
                  let traded = min !remaining maker.qty in
                  maker.qty <- maker.qty - traded;
                  remaining := !remaining - traded;
                  let ev =
                    Fill
                      {
                        f_time = time;
                        f_price = maker.price;
                        f_qty = traded;
                        f_taker_order_id = taker_id;
                        f_maker_order_id = maker.order_id;
                        f_taker_owner = taker_owner;
                        f_maker_owner = maker.owner;
                        f_taker_side = side;
                        f_pre_trade_mid = pre_trade_mid;
                      }
                  in
                  if maker.qty = 0 then begin
                    book.orders <- IntMap.remove maker.order_id book.orders;
                    walk rest (ev :: acc)
                  end
                  else (maker :: rest, ev :: acc)
          in
          let leftover, fills = walk queue [] in
          events := fills @ !events;
          match leftover with
          | [] -> set_levels book opp (IntMap.remove best (levels book opp))
          | _ -> set_levels book opp (IntMap.add best leftover (levels book opp))
        end
  done;
  (!remaining, List.rev !events)

(* ---------------------------------------------------------------- actions *)

let check_qty qty = if qty <= 0 then invalid_arg "qty must be a positive integer"

(** Submit a limit order. It first matches against any crossing liquidity as the
    taker, then rests the remainder at [price]. Returns [(order_id, events)]. *)
let submit_limit book ~owner ~side ~price ~qty ~time =
  check_qty qty;
  let order_id = new_order_id book in
  let remaining, events =
    match_incoming book ~taker_id:order_id ~taker_owner:owner ~side ~qty
      ~limit_price:(Some price) ~time
  in
  if remaining > 0 then begin
    rest_order book ~order_id ~owner ~side ~price ~qty:remaining;
    let accepted =
      Accepted
        {
          a_time = time;
          a_order_id = order_id;
          a_owner = owner;
          a_side = side;
          a_price = price;
          a_qty = remaining;
        }
    in
    (order_id, events @ [ accepted ])
  end
  else (order_id, events)

(** Submit a market order. Quantity that cannot be matched is discarded and
    reported as [Cancelled] with reason ["unfilled_market"]. *)
let submit_market book ~owner ~side ~qty ~time =
  check_qty qty;
  let order_id = new_order_id book in
  let remaining, events =
    match_incoming book ~taker_id:order_id ~taker_owner:owner ~side ~qty
      ~limit_price:None ~time
  in
  if remaining > 0 then
    let unfilled =
      Cancelled
        {
          c_time = time;
          c_order_id = order_id;
          c_owner = owner;
          c_side = side;
          c_price = None;
          c_qty = remaining;
          c_reason = "unfilled_market";
        }
    in
    (order_id, events @ [ unfilled ])
  else (order_id, events)

(** Cancel a resting order. Returns [None] if the order is unknown or already
    fully filled, mirroring real venues where cancels race with fills. *)
let cancel book ~order_id ~time =
  match IntMap.find_opt order_id book.orders with
  | None -> None
  | Some order ->
      book.orders <- IntMap.remove order_id book.orders;
      let m = levels book order.side in
      let q = IntMap.find order.price m in
      let q' = List.filter (fun o -> o.order_id <> order_id) q in
      (match q' with
      | [] -> set_levels book order.side (IntMap.remove order.price m)
      | _ -> set_levels book order.side (IntMap.add order.price q' m));
      Some
        {
          c_time = time;
          c_order_id = order_id;
          c_owner = order.owner;
          c_side = order.side;
          c_price = Some order.price;
          c_qty = order.qty;
          c_reason = "user";
        }

(* ------------------------------------------------------------- invariants *)

(** Assert internal consistency. Used by tests after every operation; never
    called on the hot path. *)
let check_invariants book =
  (match (best_bid book, best_ask book) with
  | Some b, Some a ->
      if b >= a then
        failwith (Printf.sprintf "book is crossed: bid %d >= ask %d" b a)
  | _ -> ());
  let seen = ref [] in
  List.iter
    (fun side ->
      IntMap.iter
        (fun price q ->
          if q = [] then failwith (Printf.sprintf "empty level %d left in book" price);
          let seqs = List.map (fun o -> o.entry_seq) q in
          if seqs <> List.sort compare seqs then
            failwith (Printf.sprintf "FIFO order violated at level %d" price);
          List.iter
            (fun o ->
              if o.qty <= 0 then
                failwith (Printf.sprintf "zero-qty order %d in book" o.order_id);
              if o.price <> price || o.side <> side then
                failwith (Printf.sprintf "order %d filed at wrong level" o.order_id);
              (match IntMap.find_opt o.order_id book.orders with
              | Some o' when o' == o -> ()
              | _ ->
                  failwith
                    (Printf.sprintf "order %d missing from index" o.order_id));
              seen := o.order_id :: !seen)
            q)
        (levels book side))
    [ Buy; Sell ];
  let indexed = List.map fst (IntMap.bindings book.orders) in
  if List.sort compare !seen <> List.sort compare indexed then
    failwith "order index out of sync with levels"
